### PR TITLE
lib: introduce m0_trace_level_allow() API

### LINKED ISSUE #1317: Improve API to switch m0trace completely off at startup time
Currently, there is `m0_trace_set_mmapped_buffer(false)` API which disables the creation of mmap-ped file with the tracing buffer, but it does not disable the tracing - which still might be the bottleneck and kill the concurrency when many threads are used in the process. (Note: tracing uses the atomic variable for records allocation, which is kills scalability.)

Suggestion: improve `m0_trace_set_mmapped_buffer(false)` API to switch tracing completely OFF.

### PR DESCRIPTION
**Pull request by andriytk**
# Problem Statement

`m0_trace_allot()` uses the atomic variable to allocate the trace records in the buffer, so with many
trace records generated in each motr thread of a multithreaded application, the more threads are working in parallel, the more they get hampered by the tracing subsystem. This impacts the scalability of the applications.

It can easily be seen, for example, with go/mcp utility: when using many threads - `perf top` always shows `m0_trace_allot()` at the top, and the performance does not increase after 4 threads. Moreover, it degrades with adding more threads (although the node has enough CPU cores).

# Design

Introduce m0_trace_level_allow(level) API, and return from `m0_trace_allot()` immediately if the trace level of the record is more than the set one. It will allow motr apps to specify the needed level of tracing allowed in run-time and reduce the impact of the tracing on the scalability of multithreaded apps.

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [x] Interface change (if any) are documented
- [x] Side effects on other features (deployment/upgrade)
- [x] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

Closes #1317.

### CODE REVIEW

**Comment by andriytk** [comment]
Here is `perf top` of writing motr object with `mcp -threads 8` without the patch:

image

and with the patch:

image

File: lib/trace.c [hunk 1]
```diff
@@ -129,6 +129,10 @@ static const char *trace_print_ctx_str[] = {
 	[M0_TRACE_PCTX_FULL]  = "full",
 };
 
+#ifndef __KERNEL__
+M0_INTERNAL bool m0_trace_use_mmapped_buffer(void);
+#endif
```
Comment (nikitadanilov) [inline lib/trace.c:134]: This is not entirely right. Using mmapped buffer and enabling trace should be separate options. An additional interface

m0_trace_disable(bool);
bool m0_trace_is_disabled();

(available in kernel too) is needed.

Besides, this also silently discards error messages, I don't think this is acceptable.

Wouldn't it be better to make M0_TRACE_HIGHEST_ALLOWED_LEVEL a variable that the application can adjust to suppress low priority messages?

**Review by nikitadanilov — CHANGES_REQUESTED** [review]

File: lib/trace.c (hunk 1 above)
Comment (madhavemuri) [reply lib/trace.c:134]: > Here is `perf top` of writing motr object with `mcp -threads 8` without the patch:
> 
> image
> 
> and with the patch:
> 
> image

@andriytk : Are you seeing the high utilization when only ERROR and above levels of trace are used ,
  Following config level can be set during configure
  --with-trace-max-level=<M0_ALWAYS|M0_FATAL|M0_ERROR|M0_WARN|M0_NOTICE|M0_INFO|M0_DEBUG>
                          set highest trace level that will be enabled, all
                          levels with verbosity above that will not be logged

Comment (andriytk) [reply lib/trace.c:134]: No, I didn't check that. It's very inconvenient to change this at compile time. Note: Hare also has to be re-compiled when Motr is recompiled. This is unpractical.

But I don't think that having only ERRORs will cause any issues. We don't have very much ERROR traces, especially, everything is working fine, do we?

Comment (andriytk) [reply lib/trace.c]: > Wouldn't it be better to make M0_TRACE_HIGHEST_ALLOWED_LEVEL a variable that the application can adjust to suppress low priority messages?

Agree. Implemented, see the updated code.

File: lib/trace.c [hunk 2]
```diff
@@ -213,6 +213,15 @@ static void copy_string_data(char *dst_str, const char *body,
 		}
 }
 
+static unsigned allowed_level = -1;
```
Comment (nikitadanilov) [inline lib/trace.c:216]: UINT_MAX or M0_CALL, please.

**Review by nikitadanilov — CHANGES_REQUESTED** [review]

File: lib/trace.c (hunk 2 above)
Comment (andriytk) [reply lib/trace.c]: Fixed, thanks.

**Review by nikitadanilov — APPROVED** [review]

**Review by madhavemuri — APPROVED** [review]

**Comment by andriytk** [comment]
Some more test results from the SAGE cluster: compare the performance:

`mcp` with 8 threads with tracing enabled:

```
15:35 tkachuk1@sage-tier1-1:mcp$ ./mcp -ep 172.18.1.1@o2ib:12345:4:1 -prof 0x7000000000000001:0x3a1 -proc 0x7200000000000001:0x19 -hax 172.18.1.1@o2ib:12345:1:1 -pool 0x6f00000000000001:0x2e7 -v -bsz 1024 -threads 8 -osz $((1024*1024)) -trace /dev/zero 0x3456023:0x87002806 
2021/12/16 15:36:07 mio.go:558: W: off=0 len=1073741824 bs=67108864 gs=33554432 speed=143 (Mbytes/sec)
2021/12/16 15:36:14 mio.go:558: W: off=1073741824 len=1073741824 bs=67108864 gs=33554432 speed=138 (Mbytes/sec)
2021/12/16 15:36:21 mio.go:558: W: off=2147483648 len=1073741824 bs=67108864 gs=33554432 speed=152 (Mbytes/sec)
2021/12/16 15:36:32 mio.go:558: W: off=3221225472 len=1073741824 bs=67108864 gs=33554432 speed=105 (Mbytes/sec)
2021/12/16 15:36:42 mio.go:558: W: off=4294967296 len=1073741824 bs=67108864 gs=33554432 speed=105 (Mbytes/sec)
^C
```

Without tracing (disabled):

```
15:35 tkachuk1@sage-tier1-1:mcp$ ./mcp -ep 172.18.1.1@o2ib:12345:4:1 -prof 0x7000000000000001:0x3a1 -proc 0x7200000000000001:0x19 -hax 172.18.1.1@o2ib:12345:1:1 -pool 0x6f00000000000001:0x2e7 -v -bsz 1024 -threads 8 -osz $((1024*1024)) /dev/zero 0x3456023:0x87002806 
2021/12/16 15:35:23 mio.go:558: W: off=0 len=1073741824 bs=67108864 gs=33554432 speed=163 (Mbytes/sec)
2021/12/16 15:35:29 mio.go:558: W: off=1073741824 len=1073741824 bs=67108864 gs=33554432 speed=212 (Mbytes/sec)
2021/12/16 15:35:34 mio.go:558: W: off=2147483648 len=1073741824 bs=67108864 gs=33554432 speed=214 (Mbytes/sec)
2021/12/16 15:35:38 mio.go:558: W: off=3221225472 len=1073741824 bs=67108864 gs=33554432 speed=237 (Mbytes/sec)
2021/12/16 15:35:44 mio.go:558: W: off=4294967296 len=1073741824 bs=67108864 gs=33554432 speed=204 (Mbytes/sec)
^C
```

**Comment by cortx-admin** [comment]
<h1>Jenkins CI Result : <a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964/>Motr#964</a></h1><h2>Motr Test Summary</h2><table><tr><th>Test Result</th><th>Count</th><th>Info</tr><tr><td>:x:Failed</td><td>1</td><td><details><summary>:file_folder:</summary><p><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/01motr-single-node/00userspace-tests target=_blank>01motr-single-node/00userspace-tests</a></p></details></td></tr><tr><td>:checkered_flag:Skipped</td><td>5</td><td><details><summary>:file_folder:</summary><p><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/01motr-single-node/37protocol target=_blank>01motr-single-node/37protocol</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/04motr-single-node/28sys-kvs-kernel target=_blank>04motr-single-node/28sys-kvs-kernel</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/05motr-single-node/07mount-fail target=_blank>05motr-single-node/07mount-fail</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/05motr-single-node/07mount-multiple target=_blank>05motr-single-node/07mount-multiple</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/05motr-single-node/17sns-repair-concurrent-io target=_blank>05motr-single-node/17sns-repair-concurrent-io</a></p></details></td></tr><tr><td>:heavy_check_mark:Passed</td><td>61</td><td><details><summary>:file_folder:</summary><p><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/01motr-single-node/01kernel-tests target=_blank>01motr-single-node/01kernel-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/01motr-single-node/01net target=_blank>01motr-single-node/01net</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/01motr-single-node/02rpcping target=_blank>01motr-single-node/02rpcping</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/01motr-single-node/03console target=_blank>01motr-single-node/03console</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/01motr-single-node/04initscripts target=_blank>01motr-single-node/04initscripts</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/01motr-single-node/05confgen target=_blank>01motr-single-node/05confgen</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/01motr-single-node/06hagen target=_blank>01motr-single-node/06hagen</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/01motr-single-node/28sys-kvs target=_blank>01motr-single-node/28sys-kvs</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/01motr-single-node/35m0singlenode target=_blank>01motr-single-node/35m0singlenode</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/01motr-single-node/43m0crate target=_blank>01motr-single-node/43m0crate</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/01motr-single-node/52motr-singlenode-sanity target=_blank>01motr-single-node/52motr-singlenode-sanity</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/02motr-single-node/07m0d-fatal target=_blank>02motr-single-node/07m0d-fatal</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/02motr-single-node/08spiel target=_blank>02motr-single-node/08spiel</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/02motr-single-node/10pver-assign target=_blank>02motr-single-node/10pver-assign</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/02motr-single-node/11m0t1fs target=_blank>02motr-single-node/11m0t1fs</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/02motr-single-node/13dgmode-io target=_blank>02motr-single-node/13dgmode-io</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/02motr-single-node/14poolmach target=_blank>02motr-single-node/14poolmach</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/02motr-single-node/20rpc-session-cancel target=_blank>02motr-single-node/20rpc-session-cancel</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/02motr-single-node/21fsync-single-node target=_blank>02motr-single-node/21fsync-single-node</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/02motr-single-node/26motr-user-kernel-tests target=_blank>02motr-single-node/26motr-user-kernel-tests</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/02motr-single-node/41motr-conf-update target=_blank>02motr-single-node/41motr-conf-update</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/02motr-single-node/51kem target=_blank>02motr-single-node/51kem</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/02motr-single-node/53clusterusage-alert target=_blank>02motr-single-node/53clusterusage-alert</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/02motr-single-node/67fdmi-plugin-multi-filters target=_blank>02motr-single-node/67fdmi-plugin-multi-filters</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/03motr-single-node/04sss target=_blank>03motr-single-node/04sss</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/03motr-single-node/06conf target=_blank>03motr-single-node/06conf</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/03motr-single-node/08spiel-multi-confd target=_blank>03motr-single-node/08spiel-multi-confd</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/03motr-single-node/24motr-dix-repair-lookup-insert target=_blank>03motr-single-node/24motr-dix-repair-lookup-insert</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/03motr-single-node/36spare-reservation target=_blank>03motr-single-node/36spare-reservation</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/03motr-single-node/60sns-repair-motr-1f target=_blank>03motr-single-node/60sns-repair-motr-1f</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/03motr-single-node/61sns-repair-motr-1n-1f target=_blank>03motr-single-node/61sns-repair-motr-1n-1f</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/03motr-single-node/62sns-repair-motr-mf target=_blank>03motr-single-node/62sns-repair-motr-mf</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/03motr-single-node/63sns-repair-motr-1k-1f target=_blank>03motr-single-node/63sns-repair-motr-1k-1f</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/03motr-single-node/64sns-repair-motr-ios-fail target=_blank>03motr-single-node/64sns-repair-motr-ios-fail</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/03motr-single-node/65sns-repair-motr-abort target=_blank>03motr-single-node/65sns-repair-motr-abort</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/03motr-single-node/66sns-repair-motr-abort-quiesce target=_blank>03motr-single-node/66sns-repair-motr-abort-quiesce</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/04motr-single-node/08spiel-sns-repair target=_blank>04motr-single-node/08spiel-sns-repair</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/04motr-single-node/08spiel-sns-repair-quiesce target=_blank>04motr-single-node/08spiel-sns-repair-quiesce</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/04motr-single-node/11m0t1fs-rconfc-fail target=_blank>04motr-single-node/11m0t1fs-rconfc-fail</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/04motr-single-node/19sns-repair-abort target=_blank>04motr-single-node/19sns-repair-abort</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/04motr-single-node/22sns-repair-ios-fail target=_blank>04motr-single-node/22sns-repair-ios-fail</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/04motr-single-node/25m0kv target=_blank>04motr-single-node/25m0kv</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/04motr-single-node/34sns-repair-1n-1f target=_blank>04motr-single-node/34sns-repair-1n-1f</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/04motr-single-node/44motr-rm-lock-cc-io target=_blank>04motr-single-node/44motr-rm-lock-cc-io</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/04motr-single-node/45motr-rmw target=_blank>04motr-single-node/45motr-rmw</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/04motr-single-node/48motr-raid0-io target=_blank>04motr-single-node/48motr-raid0-io</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/04motr-single-node/49motr-rpc-cancel target=_blank>04motr-single-node/49motr-rpc-cancel</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/05motr-single-node/07mount target=_blank>05motr-single-node/07mount</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/05motr-single-node/12fsync target=_blank>05motr-single-node/12fsync</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/05motr-single-node/12fwait target=_blank>05motr-single-node/12fwait</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/05motr-single-node/15sns-repair-single target=_blank>05motr-single-node/15sns-repair-single</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/05motr-single-node/16sns-repair-multi target=_blank>05motr-single-node/16sns-repair-multi</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/05motr-single-node/18sns-repair-quiesce target=_blank>05motr-single-node/18sns-repair-quiesce</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/05motr-single-node/23sns-abort-quiesce target=_blank>05motr-single-node/23sns-abort-quiesce</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/05motr-single-node/23spiel-dix-repair target=_blank>05motr-single-node/23spiel-dix-repair</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/05motr-single-node/23spiel-dix-repair-quiesce target=_blank>05motr-single-node/23spiel-dix-repair-quiesce</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/05motr-single-node/40motr-dgmode target=_blank>05motr-single-node/40motr-dgmode</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/05motr-single-node/42motr-utils target=_blank>05motr-single-node/42motr-utils</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/05motr-single-node/43motr-sync-replication target=_blank>05motr-single-node/43motr-sync-replication</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/05motr-single-node/44motr-sns-repair target=_blank>05motr-single-node/44motr-sns-repair</a></br><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit/05motr-single-node/45motr-sns-repair-N-1 target=_blank>05motr-single-node/45motr-sns-repair-N-1</a></p></details></td></tr><tr><td>Total</td><td>67</td><td><a href=http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Motr/964//testReport/junit>:link:</a></td></tr></table><h2> CppCheck Summary</h2><p> &emsp;&emsp;&emsp;Cppcheck: No new warnings found :thumbsup:</p>

**Pull request merged**